File: html/old/backup/sacor2_2009/robotmoves.java
// SACO 2nd Round 2009
// Problem: Robot Moves
// Author:  Robert Ketteringham
import java.io.*;
import java.lang.Math;

class robotmoves {
    public static void main (String [] args) throws IOException {
        BufferedReader kb=new BufferedReader (new InputStreamReader(System.in));
        System.out.print("Enter the width: ");
        int width=Integer.parseInt(kb.readLine());
        System.out.print("Enter the height: ");
        int height=Integer.parseInt(kb.readLine());
        System.out.println("Enter the initial instructions: ");
        String instructions = kb.readLine();
        int top = 0;
        int bottom = height - 1;
        int left = 0;
        int right = width - 1;
        for (int i = 0; i < instructions.length(); i++) {
            if (instructions.charAt(i) == 'd' && top < height - 1)
                top++;
            if (instructions.charAt(i) == 'd' && bottom < height - 1)
                bottom++;
            if (instructions.charAt(i) == 'u' && top > 0)
                top--;
            if (instructions.charAt(i) == 'u' && bottom > 0)
                bottom--;
            if (instructions.charAt(i) == 'r' && left < width - 1)
                left++;
            if (instructions.charAt(i) == 'r' && right < width - 1)
                right++;
            if (instructions.charAt(i) == 'l' && left > 0)
                left--;
            if (instructions.charAt(i) == 'l' && right > 0)
                right--;
        }
        if (bottom == top) //If it will always end up in the middle, then we don't need to move it to either of the sides
            bottom = 0;
        if (left == right)
            right = 0;
        int moves = Math.min(bottom, (height - top - 1)) + Math.min(right, (width - left - 1));
        System.out.println("You need to issue a further " + moves + " instructions");
    }
} 

File: html/old/backup/sacor2_2009/robotmoves.py
#!/usr/bin/env python
# SACO Second Round 2009
# Problem: Robot Moves
# Author:  Michiel Baird

M = int(raw_input("Enter the width M: "))
N = int(raw_input("Enter the height N: "))

instructions = raw_input("Enter the initial instructions: ")
# The worst starting positions are on the sides
# Simulate those 4 then take the best moves to a corner
bottom = N-1
top = 0
left = 0
right = M-1
for i in instructions:
    if i == 'd':
        top += 1
        bottom += 1
        top = min(N-1,top)       # Ignored instruction
        bottom = min(N-1,bottom) # Ignored instruction
    elif i == 'u':
        top -= 1
        bottom -= 1
        top = max(0,top)         # Ignored instruction
        bottom = max(0,bottom)   # Ignored instruction
    elif i == 'r':
        left += 1
        right += 1
        left = min(M-1,left)     # Ignored instruction
        right = min(M-1,right)   # Ignored instruction
    elif i == 'l':
        left -= 1
        right -= 1
        left = max(0,left)       # Ignored instuction
        right = max(0,right)     # Ignored instuction

# From the worst starting positions is it the best to go up or down and left or right
if top == bottom: bottom = 0 # If the worst possible is the same no vertical instructions are necessary
if left == right: right = 0      ## If the worst possible is the same no horizontal instructions are necessary
movesreq = min(bottom ,N - top - 1) + min(right ,M - left - 1)
print "You need to issue a further %d instructions." % movesreq
